pkg/server/load_endpoint.go

```diff
@@ -16,65 +16,114 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/server/status"
+	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 )
 
-// Construct a handler responsible for serving the instant values of selected
-// load metrics. These include user and system CPU time currently.
-// TODO(knz): this should probably include memory usage too somehow.
-func makeStatusLoadHandler(
-	ctx context.Context, rsr *status.RuntimeStatSampler, metricSource metricMarshaler,
-) func(http.ResponseWriter, *http.Request) {
-	cpuUserNanos := metric.NewGauge(rsr.CPUUserNS.GetMetadata())
-	cpuSysNanos := metric.NewGauge(rsr.CPUSysNS.GetMetadata())
-	cpuNowNanos := metric.NewGauge(rsr.CPUNowNS.GetMetadata())
-	registry := metric.NewRegistry()
-	registry.AddMetric(cpuUserNanos)
-	registry.AddMetric(cpuSysNanos)
-	registry.AddMetric(cpuNowNanos)
-
-	// Exporter for the CPU metrics that are provided only by the load handler.
-	exporter := metric.MakePrometheusExporter()
-	regScrape := func(pm *metric.PrometheusExporter) {
-		pm.ScrapeRegistry(registry, true)
-	}
+var (
+	// Counter to count accesses to the prometheus vars endpoint /_status/vars .
+	telemetryPrometheusLoadVars = telemetry.GetCounterOnce("monitoring.prometheus.load_vars")
+)
+
+type loadEndpoint struct {
+	// Capture the start time of the process as well as the initial User and
+	// System CPU times so all reporting can be done relative to these.
+	initTimeNanos      int64
+	initUserTimeMillis int64
+	initSysTimeMillis  int64
+
+	cpuUserNanos  *metric.Gauge
+	cpuSysNanos   *metric.Gauge
+	cpuNowNanos   *metric.Gauge
+	uptimeSeconds *metric.GaugeFloat64
+
+	registry     *metric.Registry
+	exporterLoad metric.PrometheusExporter
+	exporterVars metric.PrometheusExporter
 
+	mainMetricSource metricMarshaler
+}
+
+func newLoadEndpoint(
+	rsr *status.RuntimeStatSampler, mainMetricSource metricMarshaler,
+) (*loadEndpoint, error) {
+	initUserTimeMillis, initSysTimeMillis, err := status.GetProcCPUTime(context.Background())
+	if err != nil {
+		return nil, err
```

Comment:
nit: add an error.Wrap annotation to this error so that in the unlikely event it occurs in production we get a stack trace that points to the load endpoint initialization.